sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java

```diff
@@ -0,0 +1,254 @@
+
+        // URL website = new URL("http://www.website.com/information.asp");
+        // ReadableByteChannel rbc = Channels.newChannel(website.openStream());
+        // FileOutputStream fos = new FileOutputStream("information.html");
+        // fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
+
+        try {
+            buildDirectory(service.getContext());
+            char forwardSlashChar = '/';
+
+            // Based on the URL build the directory in which we will copy the file.
+            // kmzFilePath and kmzDirectory are final products of this processing.
+
+            Log.d(TAG, "protocol " + service.getURL().getProtocol() + " HOST " + service.getURL().getHost()
+                    + " File " + service.getURL().getFile());
+            Log.d(TAG, "File.separator " + File.separator);
+            URLConnection connection = service.getURL().openConnection();
+            String path = service.getURL().getPath();
+
+            String hostName = service.getURL().getHost();
+            if(null == hostName) {
+                hostName = "file";
+            }
+
+            File directory;
+            String fileName;
+            String targetFilePath;
+
+            // Figure out the destination directory and destination file path.
+            if (service.getURL().getProtocol().equals("file")) {
+                String file = service.getURL().getFile();
+                if (file.contains(":")) {
+                    String split[] = file.split(":");
+                    file = split[split.length - 1];
+                }
+                fileName = file.substring(file.lastIndexOf(forwardSlashChar) + 1);
+                directory = new File(kmlsRoot + File.separator + file.replace(forwardSlashChar, '.'));
+            } else {
+                directory = new File(kmlsRoot + File.separator + hostName.replace(File.separatorChar, '.') + path.replace(File.separatorChar, '.'));
+                fileName = path.substring(path.lastIndexOf(File.separatorChar) + 1);
+            }
+
+            targetFilePath = directory + File.separator + fileName;
+
+            Log.v(TAG, "File Path from URL " + path + " target directory " + directory + " target fileName " + fileName);
+
+            if(!directory.exists()) {
+                directory.mkdirs();
+            }
+            File targetFile = new File(targetFilePath);
+
+            // If target file already exists then no need to fetch it. This can happen if client application
+            // does add/remove/add or there are multiple map instances.
+            if(targetFile.exists() && (targetFile.length() > 0) && (targetFile.canRead())) {
+                Log.i(TAG, "file " + targetFilePath + " exists, no need to fetch");
+            } else {
+
+                // Clean directory just in case and then copy the file over.
+                cleanDirectory(directory);
+
+                try (InputStream input = connection.getInputStream();
```

Comment:
Take a look at this comment on line 126, NIO was ending abruptly, so decided to go the long way, it is working reliably.

// Following would be much more efficient but it is truncating the file. If we have to support
        // Large file sizes then we should investigate the code below.

        // URL website = new URL("http://www.website.com/information.asp");
        // ReadableByteChannel rbc = Channels.newChannel(website.openStream());
        // FileOutputStream fos = new FileOutputStream("information.html");
        // fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);